### ISSUE TITLE
"resources" column for persistent workers (Slurm)

### ISSUE THREAD

**Issue by pat-s** [ropensci/drake#733]
As outlined in the [manual](https://ropenscilabs.github.io/drake-manual/hpc.html#advanced-options), transient workers can make use of a `resources` column in the plan. Is this setting also picked up by persistent workers, e.g. when using SLURM? 

Currently I use `template(list = n_jobs = <value>)` in `make()` to specify the resources but I would like to have a more fine-grained level if I built targets in parallel.

The "resources column" approach is only mentioned under the "transien workers" section so I wonder if it will work ;)

**Comment by wlandau** [comment]
Good question. Right now, unfortunately, the `resources` column only applies to transient workers. At some point, if/when `clustermq` supports heterogeneous workers, we could think about supporting worker affinities to control which targets get sent to which resources. This might eventually happen through the [`workers`](https://github.com/wlandau/workers) package or maybe even https://github.com/HenrikBengtsson/future/issues/204, the former of which has resumed development.

**Comment by wlandau** [comment]
Also, glad to hear you have access a SLURM cluster now!

**Comment by pat-s** [comment]
Thanks for clarifying! 

In this regard, I just came across the following question:

Are `make(jobs = 5)` and `make(template(list(n_jobs = 5))` doing the same? Looks like they both set the `n_jobs` variable in `SBATCH` calls which triggers a "job array". If so, I am happy to make a PR to the manual to clarify this (is the source of the manual accessible?).

Until now I thought that `n_jobs` might request a minimum cpu number of a certain job to the scheduler. However, it looks like that `#SBATCH --cpus-per-task <value>` does this. The option needs to be added to the template file by the user and then propagated through the `template` argument. 

I think also that a more in-depth explanation of "SelectType" "linear vs con_res" might be valuable for users. The first one essentially blocks a whole node for one job even if its running sequentially. Ofc this is becoming more of a slurm thing here and some users might not have admin rights to change this - however, they could at least check and report to their admin.

Not sure if `drake` is the right place, maybe this information should live in _clustermq_.

**Comment by wlandau** [comment]
> Are `make(jobs = 5)` and `make(template(list(n_jobs = 5))` doing the same? Looks like they both set the `n_jobs` variable in `SBATCH` calls which triggers a "job array". If so, I am happy to make a PR to the manual to clarify this (is the source of the manual accessible?).

Interesting. I have never tried, but I guess both should theoretically spin up 5 workers. The difference is that `make(jobs = 1, template(list(n_jobs = 5))` will only make use of one of those workers, so you will not actually see any speedup from distributed computing. Maybe a PR to the manual that discourages this would be useful.

> Until now I thought that `n_jobs` might request a minimum cpu number of a certain job to the scheduler. However, it looks like that `#SBATCH --cpus-per-task <value>` does this. The option needs to be added to the template file by the user and then propagated through the `template` argument.

Yup. In general, this depends on the scheduler. I use SGE, which has a different flag for "slots" (relevant: https://github.com/mschubert/clustermq/issues/89#issuecomment-406306693).

> I think also that a more in-depth explanation of "SelectType" "linear vs con_res" might be valuable for users. The first one essentially blocks a whole node for one job even if its running sequentially. Ofc this is becoming more of a slurm thing here and some users might not have admin rights to change this - however, they could at least check and report to their admin.

It's a good point, but I think [consumable resources](https://slurm.schedmd.com/cons_res_share.html) are out of scope for both `drake` and `clustermq`. This seems to be part of the [configuration settings](https://slurm.schedmd.com/slurm.conf.html) that sys admins set when they install SLURM.

**Comment by pat-s** [comment]
> Interesting. I have never tried, but I guess both should theoretically spin up 5 workers. The difference is that `make(jobs = 1, template(list(n_jobs = 5))` will only make use of one of those workers, so you will not actually see any speedup from distributed computing. Maybe a PR to the manual that discourages this would be useful.

Yeah sure, that is a possible pitfall. I was just wondering what happens if both are supplied, i.e. if they somehow conflict in the end or which option will actually be used.

> It's a good point, but I think [consumable resources](https://slurm.schedmd.com/cons_res_share.html) are out of scope for both `drake` and `clustermq`. This seems to be part of the [configuration settings](https://slurm.schedmd.com/slurm.conf.html) that sys admins choose when they install SLURM.

I see why they are out of scope here. However, while most user will not have a chance to change the Slurm settings of the HPC they are using, I see the following problem:
`drake` users (who are not by default deep into Slurm or HPC) will just use the default template without worrying too much as long as it works OOB.
However, often users might have _within-target parallelism_ and if `cpus-per-task` is not set in the template, this might lead to really bad outcomes, sometimes even for multiple slurm users.

### Examples

- If the `SelectType` argument of Slurm is configured as `con_res`, tasks will be distributed by cpu. If `cpus-per-task` is not explicitly set, Slurm assumes that the task will need 1 core. If within-target parallelism kicks in, actually X CPUs are required which will actually conflict with possible other tasks running on this node. It might even happen that the node will run into memory problems suddenly because it accepts more resources than it can actually handle.

- On the other hand, if `SelectType` is set to `linear`, all tasks will request a whole node, even if they are running on only one CPU. This is very inefficient. Of course, this should be handled by the Sys admin. However, sometimes Sys admins also do not have a deeper understanding of Slurm (especially in a scientific setting) and "just" install it with the defaults. Since this option is the default, I bet there are many cases out there in which this happens this way right now.

### Solution

Add `cpu-per-task` option to the default slurm template. If it is missing, the process will error and request the setting via the `template` argument.
Also it requests the user to actually check which `SelectType` argument is configured for Slurm on their system and even ask the Sys admin to eventually change it.
(FYI, `cpus-per-task` is only valid for `SelectType = con_res`.)
Even though this seems out of scope in the first place, I think this information might be vary valuable for a lot of users that would like to use HPC through `drake`. Also Slurm is AFAIK the most used scheduler used on HPCs.

I am happy to make a PR for the manual (to give a little back to you). Probably this content would also be valuable for `clustermq` as the same problems ofc also apply there. I know you already declined kindly but maybe this additional information brings some new light in. Let me know what you think.

cc @mschubert

**Comment by wlandau** [comment]
@pat-s, you bring up an important pitfall I had not considered. I would really appreciate a PR to update the [within-target parallelism section](https://github.com/ropenscilabs/drake-manual/blob/master/hpc.Rmd#L266), maybe also the [chapter of cautionary notes](https://github.com/ropenscilabs/drake-manual/blob/master/caution.Rmd) (which probably needs cleaning up at some point). I do not think we need the full details, just a warning about what could happen if you attempt to request more CPUs than the scheduler is ready to supply.

**Comment by mschubert** [comment]
Couple of comments:

> However, often users might have _within-target parallelism_ and if `cpus-per-task` is not set in the template, this might lead to really bad outcomes, sometimes even for multiple slurm users.

*clustermq* can handle this for homogeneous tasks by requesting not only the number of jobs, but also the number of cores per job.

> Add `cpu-per-task` option to the default slurm template. If it is missing, the process will error and request the setting via the `template` argument.

I should probably add this as a resource in the default *clustermq* templates: https://github.com/mschubert/clustermq/issues/123

> if/when `clustermq` supports heterogeneous workers, we could think about supporting worker affinities to control which targets get sent to which resources

I'm planning to support heterogeneous workers [via the API](https://github.com/mschubert/clustermq/wiki/Worker-API), but not via the *clustermq* package itself. 

They will always be less efficient than their homogeneous counterparts (e.g., separate submission scripts), and we will need a whole other logic to distribute tasks according to how many CPUs they use, maybe put multiple single-CPU tasks on a multi-CPU worker, etc. I'm afraid this will not happen any time soon.

> This might eventually happen through the [`workers`](https://github.com/wlandau/workers) package or maybe even [HenrikBengtsson/future#204](https://github.com/HenrikBengtsson/future/issues/204)

As I currently understand, the `future` package (even with *clustermq* support) will not be able to solve this problem.

**Comment by pat-s** [comment]
> I would really appreciate a PR to update the [within-target parallelism section](https://github.com/ropenscilabs/drake-manual/blob/master/hpc.Rmd#L266),

It's on my list :+1:
